### PR TITLE
Improvements to BiCGstab-L

### PR DESCRIPTION
**Pull request by weinbe2**
This PR includes various performance and quality-of-life improvements to the BiCGstab-L solver in QUDA. It also includes a bugfix and quality-of-life improvement for CA-CG. It also fixes a bug in `getStencilSteps` that broke the auxiliary worker class when a `direct` operator was passed in (which has existed since the accessor re-write, so clearly it hasn't bothered anyone). New options have been exposed in the MILC HISQ MG interface.

More detail:
* The MR portion of BiCGstab-L has been fused into one massive block dot product, a dense LDL decomposition+solve, and two vectorized `caxpy` calls following the definitions in the BiCGstab-L paper. This improves performance at generally no lost in accuracy (empirically).
* BiCGstab-L has been optimized for the fixed iteration case, boosting performance when used as a smoother or coarsest level solver in MG
* BiCGstab-L now supports SVD deflation.
* Both BiCGstab-L and CA-CGN(E/R) are exposed as smoothers and coarsest level solvers in the MILC HISQ MG interface.
* QoL: individual intermediate fields in BiCGstab-L are wrapped in `std::unique_ptr`
* QoL: the scalars in BiCGstab-L (and CA-CG) are now wrapped in `std::vector`s
* Bugfix: `getStencilSteps` correctly reflects that (most) non-preconditioned full-parity operators are now implemented as a single `dslash` as opposed to separate `D_eo` and `D_oe` calls.
* Bugfix: the `fixed_iteration` codepath of CA-CG computed too large of a block dot product, which as now been fixed.

### Reference fast compile command

Non-MG. Note `QUDA_PRECISION` is set to `6` instead of `4` to test mixed precision.

```
cmake -DQUDA_DIRAC_DEFAULT_OFF=ON -DQUDA_DIRAC_STAGGERED=ON -DQUDA_PRECISION=6 -DQUDA_RECONSTRUCT=4 -DQUDA_FAST_COMPILE_DSLASH=ON -DQUDA_FAST_COMPILE_REDUCE=ON ../quda
```

MG, including USQCD pieces to use physical gauge fields

```
cmake -DQUDA_DIRAC_DEFAULT_OFF=ON -DQUDA_DIRAC_STAGGERED=ON -DQUDA_PRECISION=4 -DQUDA_RECONSTRUCT=4 -DQUDA_FAST_COMPILE_DSLASH=ON -DQUDA_FAST_COMPILE_REDUCE=ON -DQUDA_MULTIGRID=ON -DQUDA_DOWNLOAD_USQCD=ON -DQUDA_QIO=ON -DQUDA_QMP=ON ../quda
```

### Tests

Simple staggered solve using the non-Schur operator using BiCGstab-4:

```
./staggered_invert_test \
        --prec single --prec-sloppy half  \
        --mass 0.1 --recon 18 --recon-sloppy 18 \
        --dim 16 16 16 16 --gridsize 1 1 1 1 --partition 0 \
        --dslash-type staggered --compute-fat-long true --tadpole-coeff 0.905160183 --tol 1e-5 \
        --verbosity verbose --inv-type bicgstab-l --ngcrkrylov 4 --test 0 \
        --nsrc 1 --niter 100
```

Simple staggered solve with the Schur operator using CA-CG(4) w/a Chebyshev basis

```
./staggered_invert_test \
        --prec single --prec-sloppy half  \
        --mass 0.1 --recon 18 --recon-sloppy 18 \
        --dim 16 16 16 16 --gridsize 1 1 1 1 --partition 0 \
        --dslash-type staggered --compute-fat-long true --tadpole-coeff 0.905160183 --tol 1e-5 \
        --verbosity verbose --inv-type ca-cg --ngcrkrylov 4 --ca-basis-type chebyshev --test 3 \
        --nsrc 1 --niter 100
```

MG solve using CA-CGNR as a smoother and BiCGstab-L as the coarsest level solve (I recommend using a "real" config for this one):

```
./staggered_invert_test \
            --prec double --prec-sloppy single --prec-null single --prec-precondition single \
            --mass 0.01 --recon 18 --recon-sloppy 18 \
            --dim 16 16 16 16 --gridsize 1 1 1 1 --load-gauge l16t16b7p0 --partition 0 \
            --dslash-type asqtad --compute-fat-long true --tadpole-coeff 0.905160183 --tol 1e-10 \
            --verbosity verbose --solve-type direct --solution-type mat --inv-type gcr \
            --inv-multigrid true --mg-levels 3 --mg-staggered-coarsen-type kd-optimized \
            --mg-block-size 0 1 1 1 1 --mg-nvec 0 3 \
            --mg-block-size 1 4 4 4 4 --mg-nvec 1 64 \
            --mg-setup-tol 1 1e-5 --mg-setup-inv 1 ca-cgnr --mg-setup-ca-basis-type 1 chebyshev --mg-setup-ca-basis-size 1 8 \
            --nsrc 1 --niter 25 \
            --mg-use-mma true \
            --mg-smoother 0 ca-cgnr --mg-smoother-solve-type 0 direct    --mg-nu-pre 0 0 --mg-nu-post 0 4 \
            --mg-smoother 1 ca-cgnr --mg-smoother-solve-type 1 direct --mg-nu-pre 1 0 --mg-nu-post 1 4 \
            --mg-coarse-solver 1 gcr --mg-coarse-solve-type 1 direct --mg-coarse-solver-tol 1 0.35 --mg-coarse-solver-maxiter 1 10 \
            --mg-coarse-solver 2 bicgstab-l --mg-coarse-solve-type 2 direct-pc --mg-coarse-solver-tol 2 0.01 --mg-coarse-solver-maxiter 2 8 --mg-coarse-solver-ca-basis-size 2 8 \
            --mg-verbosity 0 verbose --mg-verbosity 1 verbose --mg-verbosity 2 verbose
```

### Outstanding work

- [x] Updating documentation on the wiki as appropriate
- [x] `clang-format`

### CODE REVIEW

File: lib/milc_interface.cpp [hunk 1]
```diff
@@ -1295,6 +1295,7 @@ struct mgInputStruct {
   QudaInverterType setup_inv[QUDA_MAX_MG_LEVEL]; // ignored on first and last level
   double setup_tol[QUDA_MAX_MG_LEVEL];           // ignored on first and last level
   double setup_maxiter[QUDA_MAX_MG_LEVEL];       // ignored on first and last level
+  int setup_ca_basis_size[QUDA_MAX_MG_LEVEL];    // ignored on first and last level
```
Comment (maddyscientist) [inline lib/milc_interface.cpp:1306]: Perhaps this comment should be clarified that these options are ignored on the first when using the KD-specific coarsening.

File: lib/inv_bicgstabl_quda.cpp [hunk 2]
```diff
@@ -533,86 +720,35 @@ namespace quda {
...
       } // End BiCG part.
 
-      // MR part. Really just modified Gram-Schmidt.
-      // The algorithm uses the byproducts of the Gram-Schmidt to update x
-      //   and other such niceties. One day I'll read the paper more closely.
-      // Can take this from 'orthoDir' in inv_gcr_quda.cpp, hard code pipelining up to l = 8.
-      for (int j = 1; j <= n_krylov; j++) {
+      // Perform the MR portion of BiCGstab-L
+      // if we're doing a fixed number of iterations, we only need to update x
+      computeMR(x_sloppy, fixed_iteration);
```
Comment (maddyscientist) [inline lib/inv_bicgstabl_quda.cpp:698]: It should be made more obvious, e.g., wrapped in an `#ifdef` and commented, how to disable the new communication-avoiding path and enable the old path here.

File: lib/inv_bicgstabl_quda.cpp [hunk 3]
```diff
@@ -11,144 +11,287 @@
 #include <dslash_quda.h>
```
Comment (maddyscientist) [inline lib/inv_bicgstabl_quda.cpp:11]: This file could also do with some cleanup to remove the dead code, e.g. lines 166-173, 
225-227, 271-276, 686-689, 711-714.

**Comment by maddyscientist** [comment]
This PR is a nice addition to QUDA.  I've done some stability testing with this, and while it seems that there is sometimes a loss in stability for poorly conditioned systems, leading to a small increase in iteration count, the overall effect is a net time reduction, making BiCGStab(L), more often than not, more stable _and_ faster than vanilla BiCGStab.

I note further improvement in algorithm stability may be possible with a more developed mixed-precision approach.

Will approve this PR once the minor comments have been addressed.  Thanks @weinbe2 😄

File: lib/milc_interface.cpp (hunk 1 above)
Comment (weinbe2) [reply lib/milc_interface.cpp:1306]: https://github.com/lattice/quda/commit/c0b3896cf97ec6732b807087de0aea5af5e2ecd1

File: lib/inv_bicgstabl_quda.cpp (hunk 3 above)
Comment (weinbe2) [reply lib/inv_bicgstabl_quda.cpp:11]: https://github.com/lattice/quda/commit/7dfe43652b77b3310e67d99b4b85d51d4b8194bf

File: lib/inv_bicgstabl_quda.cpp (hunk 2 above)
Comment (weinbe2) [reply lib/inv_bicgstabl_quda.cpp:698]: https://github.com/lattice/quda/commit/4ca7a56414aba75a68ece64632efa50abd4a41e4

**Review by maddyscientist — APPROVED** [review]

**Pull request merged**